src/pages/GapsPatternsPage.tsx

```diff
@@ -0,0 +1,313 @@
+import React, { useContext, useEffect, useState } from 'react'
+import PropTypes from 'prop-types'
+import { getGapsAsync } from '../api/gapsService'
+import { Moment } from 'moment'
+import { DatePicker, Spin } from 'antd'
+import moment from 'moment/moment'
+import { UseDate } from './components/DateTimePicker'
+import { PageContainer } from './components/PageContainer'
+import { Row } from './components/Row'
+import { Label } from './components/Label'
+import { TEXTS } from '../resources/texts'
+import OperatorSelector from './components/OperatorSelector'
+import LineNumberSelector from './components/LineSelector'
+import { NotFound } from './components/NotFound'
+import RouteSelector from './components/RouteSelector'
+import { SearchContext } from '../model/pageState'
+import { getRoutesAsync } from '../api/gtfsService'
+import {
+  Bar,
+  BarChart,
+  CartesianGrid,
+  Legend,
+  Line,
+  Tooltip,
+  XAxis,
+  YAxis,
+  ComposedChart,
+  Cell,
+  TooltipProps,
+} from 'recharts'
+import { FormControlLabel, Radio, RadioGroup, Switch } from '@mui/material'
+
+// Define prop types for the component
+interface BusLineStatisticsProps {
+  lineRef: number
+  operatorRef: string
+  fromDate: Moment
+  toDate: Moment
+}
+
+interface HourlyData {
+  planned_hour: string
+  actual_rides: number
+  planned_rides: number
+}
+
+const now = moment()
+
+function GapsByHour({ lineRef, operatorRef, fromDate, toDate }: BusLineStatisticsProps) {
+  const [hourlyData, setHourlyData] = useState<HourlyData[]>([])
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const gapsList = await getGapsAsync(fromDate, toDate, operatorRef, lineRef)
+
+        // Convert gapsList data into hourly mapping as needed
+        const hourlyMapping: Record<string, { planned_rides: number; actual_rides: number }> = {}
+
+        for (const ride of gapsList) {
+          if (ride.gtfsTime === null) {
+            continue
+          }
+          const plannedHour = ride.gtfsTime.format('HH:mm')
+
+          if (!hourlyMapping[plannedHour]) {
+            hourlyMapping[plannedHour] = { planned_rides: 0, actual_rides: 0 }
+          }
+
+          hourlyMapping[plannedHour].planned_rides += 1
+          if (ride.siriTime) {
+            hourlyMapping[plannedHour].actual_rides += 1
+          }
+        }
+
+        const result: HourlyData[] = Object.entries(hourlyMapping).map(([hour, data]) => ({
+          planned_hour: hour,
+          actual_rides: data.actual_rides,
+          planned_rides: data.planned_rides,
+        }))
+
+        result.sort((a, b) => a.planned_hour.localeCompare(b.planned_hour))
+        setHourlyData(result)
+      } catch (error) {
+        console.error('Error fetching data:', error)
+      }
+    }
+
+    fetchData()
+  }, [lineRef, operatorRef, fromDate, toDate])
```

Comment:
let's do a `gapsListHook` tomorrow

Reply:
tomorrow is in 2 weeks for me so skipping this one for now cause I'm not sure if you just meant to export this to a func of something else :)